runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -83,6 +83,62 @@ TR::TreeLowering::lowerValueTypeOperations(TR::Node* node, TR::TreeTop* tt)
       }
    }
 
+/**
+ * @brief Copy register dependencies between GlRegDeps node at exit points.
+ *
+ * This function is only intended to work with GlRegDeps nodes for exit points,
+ * (i.e. BBEnd, branch, or jump nodes) within the same extended basic block.
+ *
+ * Register dependencies are copied "logically", meaning that the actual node
+ * used to represent a dependency won't necessarily be copied. If the reg dep
+ * is represented by a PassThrough, then the node itself is copied and its
+ * child is commoned (so it's lifetime is exteneded; note that in correctly-formed
+ * IL, the child must also be the child of a reg store in the containing block).
+ * Otherwise, the dependency must be represented by a reg load, which must have
+ * come from the GlRegDeps node at the entry point and *must* be commoned
+ * (so it won't get copied).
+ *
+ * In addition, this function allows *one* register dependency to be changed
+ * (substituted). That is, if a register dependency is found under `sourceNode`
+ * for the same register that is set on `subsituteNode`, then `substituteNode`
+ * will be used instead of the dependency from `sourceNode`. Note that the
+ * reference of of `substituteNode` is incremented if/when it gets added. If
+ * `substituteNode` is NULL the no substitution will be attempted.
+ *
+ * @param targetNode is the GlRegDeps node that reg deps are copied to
+ * @param sourceNode is the GlRegDeps node that reg deps are copied from
+ * @param substituteNode is the reg dep node to substitute if a matching register is found in `sourceNode` (NULL if none)
+ */
+void
+copyExitRegDepsAndSubstitue(TR::Node* const targetNode, TR::Node* const sourceNode, TR::Node* const substituteNode)
+   {
+   for (int i = 0; i < sourceNode->getNumChildren(); ++i)
+      {
+      TR::Node* child = sourceNode->getChild(i);
+      if (substituteNode
+          && child->getLowGlobalRegisterNumber() == substituteNode->getLowGlobalRegisterNumber()
+          && child->getHighGlobalRegisterNumber() == substituteNode->getHighGlobalRegisterNumber())
+         targetNode->setAndIncChild(i, substituteNode);
+      else if (child->getOpCodeValue() == TR::PassThrough)
+         {
+         // PassThrough nodes cannot be commoned because doing so does not
+         // actually anchor the child, causing it's lifetime to not be extended
+         child = TR::Node::copy(child);
+         if (child->getFirstChild())
+            {
+            child->getFirstChild()->incReferenceCount();
+            }
+         child->setReferenceCount(1);
+         targetNode->setChild(i, child);
```

Comment:
No, because the ref count of the original child is arbitrary. `TR::Node::copy()` will just copy over the ref count as is. So we need to force the ref count of the copy to be 1. You could use:

```
child->setReferenceCount(0);
targetNode->setAndIncChild(i, child);
```

But, you still have to force the ref count to a known value. Otherwise, you increment whatever the ref count of the original node is, which is definitely wrong.